server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,392 @@
+
+instance ( J.ToJSON (Param p T.Text)
+         , J.ToJSON (Param p J.Value)
+         , J.ToJSON (Param p Int)
+         , J.ToJSON (Param p RetryConfST)
+         ) =>
+         J.ToJSON (ScheduledEvent p) where
+  toJSON = $(J.mkToJSON (J.aesonDrop 2 J.snakeCase) {J.omitNothingFields = True} ''ScheduledEvent)
+
+runScheduledEventsGenerator ::
+     L.Logger L.Hasura
+  -> Q.PGPool
+  -> IO SchemaCache
+  -> IO ()
+runScheduledEventsGenerator logger pgpool getSC = do
+  forever $ do
+    traceM "entering scheduled events generator"
+    sc <- getSC
+    let scheduledTriggers = Map.elems $ scScheduledTriggers sc
+    runExceptT
+      (Q.runTx
+         pgpool
+         (Q.ReadCommitted, Just Q.ReadWrite)
+         (insertScheduledEventsFor scheduledTriggers) ) >>= \case
+      Right _ -> pure ()
+      Left err ->
+        L.unLogger logger $ EventInternalErr $ err500 Unexpected (T.pack $ show err)
+    threadDelay oneHour
+
+insertScheduledEventsFor :: [ScheduledTriggerInfo] -> Q.TxE QErr ()
+insertScheduledEventsFor scheduledTriggers = do
+  currentTime <- liftIO getCurrentTime
+  let scheduledEvents = concatMap (generateScheduledEventsFrom currentTime) scheduledTriggers
+  case scheduledEvents of
+    []     -> pure ()
+    events -> do
+      let insertScheduledEventsSql = TB.run $ toSQL
+            SQLInsert
+              { siTable    = scheduledEventsTable
+              , siCols     = map (PGCol . T.pack) ["name", "scheduled_time"]
+              , siValues   = ValuesExp $ map (toTupleExp . toArr) events
+              , siConflict = Just $ DoNothing Nothing
+              , siRet      = Nothing
+              }
+      Q.unitQE defaultTxErrorHandler (Q.fromText insertScheduledEventsSql) () False
+  where
+    toArr (ScheduledEventSeed n t) = [(triggerNameToTxt n), (formatTime' t)]
+    toTupleExp = TupleExp . map SELit
+
+generateScheduledEventsFrom :: UTCTime -> ScheduledTriggerInfo-> [ScheduledEventSeed]
+generateScheduledEventsFrom time ScheduledTriggerInfo{..} =
+  let events =
+        case stiSchedule of
+          OneOff _ -> empty -- one-off scheduled events are generated during creation
+          Cron cron ->
+            generateSchedulesBetween
+              time
+              (addUTCTime nominalDay time)
+              cron
+   in map (ScheduledEventSeed stiName) events
+
+-- generates events (from, till] according to CronSchedule
+generateSchedulesBetween :: UTCTime -> UTCTime -> CronSchedule -> [UTCTime]
+generateSchedulesBetween from till cron = takeWhile ((>=) till) $ go from
+  where
+    go init =
+      case nextMatch cron init of
+        Nothing   -> []
+        Just next -> next : (go next)
+
+processScheduledQueue
+  :: HasVersion
+  => L.Logger L.Hasura
+  -> Q.PGPool
+  -> HTTP.Manager
+  -> IO SchemaCache
+  -> IO ()
+processScheduledQueue logger pgpool httpMgr getSC =
+  forever $ do
+    traceM "entering processor queue"
+    sc <- getSC
+    let scheduledTriggersInfo = scScheduledTriggers sc
```

Comment:
It looks like `sc` is only used right here, so you might as well write `scheduledTriggersInfo  <- scScheduledTriggers <$> getSC`.